<html xmlns:o="urn:schemas-microsoft-com:office:office"xmlns:w="urn:schemas-microsoft-com:office:word"xmlns="http://www.w3.org/TR/REC-html40"><head><meta name=Title content="What is CoGe"><meta http-equiv=Content-Type content="text/html; charset=macintosh"><link rel="stylesheet" type="text/css" href="/CoGe/css/coge.css" /><title>What is CoGe</title><style><!-- /* Font Definitions */@font-face	{font-family:"Times New Roman";	panose-1:0 2 2 6 3 5 4 5 2 3;	mso-font-charset:0;	mso-generic-font-family:auto;	mso-font-pitch:variable;	mso-font-signature:50331648 0 0 0 1 0;} /* Style Definitions */p.MsoNormal, li.MsoNormal, div.MsoNormal	{mso-style-parent:"";	margin:0in;	margin-bottom:.0001pt;	mso-pagination:widow-orphan;	font-size:12.0pt;	font-family:"Times New Roman";}table.MsoNormalTable	{mso-style-parent:"";	font-size:10.0pt;	font-family:"Times New Roman";}@page Section1	{size:8.5in 11.0in;	margin:1.0in 1.25in 1.0in 1.25in;	mso-header-margin:.5in;	mso-footer-margin:.5in;	mso-paper-source:0;}div.Section1	{page:Section1;}--></style></head><body bgcolor=white lang=EN-US style='tab-interval:.5in'><TABLE> <TR>  <TD><a href="/CoGe" border=0><img src='/CoGe/picts/CoGe-logo.png' border=0></a>  <TD id="title"><font class="title1">CoGe Faq</font></td> </TR></TABLE><table width="100%" border="0" cellspacing="0" cellpadding="0" border=0> <tr>   <td width="6"><img src="/CoGe/picts/table/table_bg_head_tl.png" width="6" height="23"></td>  <td width="6"  background="/CoGe/picts/table/table_bg_head_t.png">&nbsp;</td>  <td width="100%"  background="/CoGe/picts/table/table_bg_head_t.png"><font class="oblique">Welcome!</font></td>  <td width="6"  background="/CoGe/picts/table/table_bg_head_t.png">&nbsp;</td>  <td width="6"><img src="/CoGe/picts/table/table_bg_head_tr.png" width="6" height="23"></td> </tr> <tr>   <td height="9" valign="top" background="/CoGe/picts/table/table_bg_head_l.png"><img src="/CoGe/picts/table/spacer-1x1.png"></td>  <td height="9" valign="top"><img src="/CoGe/picts/table/table_bg_body_tl.png" width="6" height="9"></td>  <td height="9" valign="top" background="/CoGe/picts/table/table_bg_body_t.png"><img src="/CoGe/picts/table/spacer-1x1.png"></td>  <td height="9" valign="top"><img src="/CoGe/picts/table/table_bg_body_tr.png" width="6" height="9"></td>  <td height="9" valign="top" width="6" background="/CoGe/picts/table/table_bg_head_r.png"><img src="/CoGe/picts/table/spacer-1x1.png"></td> </tr> <tr>  <td background="/CoGe/picts/table/table_bg_head_l.png">&nbsp;</td>  <td background="/CoGe/picts/table/table_bg_body_l.png">&nbsp;</td>  <td> <!-- ########################################## --><!-- Main body of the central table goes in here--><!-- ########################################## --><div class=Section1><p class=MsoNormal style='mso-pagination:none;mso-layout-grid-align:none;text-autospace:none'><span style='font-family:Helvetica'><b>What is CoGe?<o:p></o:p></b></span></p><p class=MsoNormal style='mso-pagination:none;mso-layout-grid-align:none;text-autospace:none'><span style='font-family:Helvetica'><![if !supportEmptyParas]>&nbsp;<![endif]><o:p></o:p></span></p><p class=MsoNormal style='mso-pagination:none;mso-layout-grid-align:none;text-autospace:none'><span style='font-family:Helvetica'>CoGe's name is derivedfrom the phrase "Comparative Genomics" which is the field of computationalbiology that is centered on the discovery of patterns in genomes throughintergenomic comparisons.<span style="mso-spacerun: yes">&nbsp; </span>Thesepatterns are interesting because they reflect the function or evolutionaryhistory of genomes or genomic regions.<span style="mso-spacerun: yes">&nbsp;</span>Examples of such genome patterns are:<o:p></o:p></span></p><p class=MsoNormal style='mso-pagination:none;mso-layout-grid-align:none;text-autospace:none'><span style='font-family:Helvetica'>1. Series oforthologous/paralogous genes that have a conservation of order in agenome.<span style="mso-spacerun: yes">&nbsp; </span>Two or more genomicregions (either within the same organism or between different organisms) thathave a conservation of gene order is indicative that the genomic regions shareda common ancestral genomic region.<span style="mso-spacerun: yes">&nbsp;</span>Such regions are called syntenic and in the case of intragenomicdetection, are used to determine if an organism experienced a genomeduplication event.<o:p></o:p></span></p><p class=MsoNormal style='mso-pagination:none;mso-layout-grid-align:none;text-autospace:none'><span style='font-family:Helvetica'>2. ConservedNon-coding Sequences around orthologous/paralogous genes.<spanstyle="mso-spacerun: yes">&nbsp; </span>These sequences are presumed to beunder evolutionary selective constraints because they have a function (probablyinvolved with the regulation of gene expression either directly through therecruitment of transcription accessory factors/proteins or indirectly throughchomatin remodeling.)<o:p></o:p></span></p><p class=MsoNormal style='mso-pagination:none;mso-layout-grid-align:none;text-autospace:none'><span style='font-family:Helvetica'><![if !supportEmptyParas]>&nbsp;<![endif]><o:p></o:p></span></p><p class=MsoNormal style='mso-pagination:none;mso-layout-grid-align:none;text-autospace:none'><span style='font-family:Helvetica'>CoGe was designed inorder to simplify access to genomic information and the development ofcomputational tools for use in comparative genomics.<o:p></o:p></span></p><p class=MsoNormal style='mso-pagination:none;mso-layout-grid-align:none;text-autospace:none'><span style='font-family:Helvetica'><![if !supportEmptyParas]>&nbsp;<![endif]><o:p></o:p></span></p><p class=MsoNormal style='mso-pagination:none;mso-layout-grid-align:none;text-autospace:none'><span style='font-family:Helvetica'><b>There are alreadymany existing databases and tools for accessing genomic information.<spanstyle="mso-spacerun: yes">&nbsp; </span>Why make another set of resources?<o:p></o:p></b></span></p><p class=MsoNormal style='mso-pagination:none;mso-layout-grid-align:none;text-autospace:none'><span style='font-family:Helvetica'><b><![if !supportEmptyParas]>&nbsp;<![endif]><o:p></o:p></b></span></p><p class=MsoNormal style='mso-pagination:none;mso-layout-grid-align:none;text-autospace:none'><span style='font-family:Helvetica'>A current limitationof many genome databases is their organism centric organization.<spanstyle="mso-spacerun: yes">&nbsp; </span>For example, Ensembl's genome databaseis designed to store a single organism's genome; thus requires a separateinstallation of their database for each genome.<span style="mso-spacerun:yes">&nbsp; </span>This approach limits the tractability and maintainability ofgenomic information as each new genome or new release of an existing genomerequires a separate database installation.<span style="mso-spacerun:yes">&nbsp; </span>Likewise, this "one genome&#8211;one database" approach alsolimits the development and deployment of new comparative genomics tools.<spanstyle="mso-spacerun: yes">&nbsp; </span>If new tools or data organizationschemas are developed for one genome, redeploying them for previously existinggenomes requires patching each installation separately.<spanstyle="mso-spacerun: yes">&nbsp; </span>Here we propose to develop a newdatabase and application programming interface (API) to store any genome in oneunified system.<o:p></o:p></span></p><p class=MsoNormal style='mso-pagination:none;mso-layout-grid-align:none;text-autospace:none'><span style='font-family:Helvetica'><![if !supportEmptyParas]>&nbsp;<![endif]><o:p></o:p></span></p><p class=MsoNormal style='mso-pagination:none;mso-layout-grid-align:none;text-autospace:none'><span style='font-family:Helvetica'>The principalobjective of this effort is to design a relational database schema that allowsfor the tracking of any genome in an organism non-specific manner that alsoallows for the co-existence of multiple versions of the same genome. <spanstyle="mso-spacerun: yes">&nbsp;</span>By have a unified database for allgenomes, any tool developed for analyzing one organism's genome is immediatelyavailable to all genomes in the database.<span style="mso-spacerun: yes">&nbsp;</span>This approach also makes the development of comparative genomic toolseasier by having all genomic information in one centralized location, andsimilarly, if comparative genomic tools &shy;are developed to analyze one setof genomes, they are immediately available to any set of genomes.<o:p></o:p></span></p><p class=MsoNormal style='mso-pagination:none;mso-layout-grid-align:none;text-autospace:none'><span style='font-family:Helvetica'><![if !supportEmptyParas]>&nbsp;<![endif]><o:p></o:p></span></p><p class=MsoNormal style='mso-pagination:none;mso-layout-grid-align:none;text-autospace:none'><span style='font-family:Helvetica'><b>What is in CoGesystem?</b></span><span style='font-family:Helvetica'><o:p></o:p></span></p><p class=MsoNormal style='mso-pagination:none;mso-layout-grid-align:none;text-autospace:none'><span style='font-family:Helvetica'><![if !supportEmptyParas]>&nbsp;<![endif]><o:p></o:p></span></p><p class=MsoNormal style='mso-pagination:none;mso-layout-grid-align:none;text-autospace:none'><span style='font-family:Helvetica'>CoGe is a collectionof software resources that includes a database, visualization packages, andanalysis tools.<span style="mso-spacerun: yes">&nbsp; </span>These items can beused to create software tools to access and analyze genomic information in anynumber of ways.<span style="mso-spacerun: yes">&nbsp; </span>Tools that wedevelop tend to fall into three general categories:<span style="mso-spacerun:yes">&nbsp; </span>APIs, command-line tools, and web-based tools.<spanstyle="mso-spacerun: yes">&nbsp; </span>The APIs (application programminginterfaces) are an on-going collection of libraries used to access genomicinformation in the CoGe database, generate visual representation of genomicinformation, and give access to other informatics tools and algorithmsdeveloped in-house.<span style="mso-spacerun: yes">&nbsp; </span>Command-linetools usually center around data import and export from the genomesdatabase.<span style="mso-spacerun: yes">&nbsp; </span>Web-based tools utilizeweb 2.0 design principles to make interactive and easy&#8211;to-use comparativegenomic analysis and visualization tools.<span style="mso-spacerun: yes">&nbsp;</span>These tools are made available for general use on the CoGe web-site.<o:p></o:p></span></p><p class=MsoNormal style='mso-pagination:none;mso-layout-grid-align:none;text-autospace:none'><span style='font-family:Helvetica'><![if !supportEmptyParas]>&nbsp;<![endif]><o:p></o:p></span></p><p class=MsoNormal style='mso-pagination:none;mso-layout-grid-align:none;text-autospace:none'><span style='font-family:Helvetica'><b>What softwaremodels, databases, and languages are used by CoGe?</b></span><spanstyle='font-family:Helvetica'><o:p></o:p></span></p><p class=MsoNormal style='mso-pagination:none;mso-layout-grid-align:none;text-autospace:none'><span style='font-family:Helvetica'><![if !supportEmptyParas]>&nbsp;<![endif]><o:p></o:p></span></p><p class=MsoNormal><span style='font-family:Helvetica'>The CoGe database iswritten for MySQL.<span style="mso-spacerun: yes">&nbsp; </span>Its designfocuses on minimizing data redundancy and maximizing data retrieval speedswhile maintaining simplicity of data structures and relationships.<spanstyle="mso-spacerun: yes">&nbsp; </span>The core of the CoGe database centerson the storage of any genome, paying attention to the need to accommodategenomes for different sources (e.g. genbank, TIGR, JGI, Ensembl), multipledatasets for a given genome (e.g separate datasets for each chromosome,additional annotation information for a given genome), and accommodatingmultiple versions of a given genome (although data quality should increase as afunction of time, analyses performed on old datasets require old the olddatasets for subsequent validation efforts).<o:p></o:p></span></p><p class=MsoNormal><span style='font-family:Helvetica'><![if !supportEmptyParas]>&nbsp;<![endif]><o:p></o:p></span></p><p class=MsoNormal><span style='font-family:Helvetica'>APIs, components ofweb-tools, and most command-line tools are written in PERL.<spanstyle="mso-spacerun: yes">&nbsp; </span>PERL's near ubiquitous presence in web applicationsand bioinformatics, extensive additional libraries for performing web-based anddatabase transactions, and large user base makes it an ideal language forCoGe.<span style="mso-spacerun: yes">&nbsp; </span>We use PERL's objectoriented programming style extensively as well as POD for API documentation.<o:p></o:p></span></p><p class=MsoNormal><span style='font-family:Helvetica'><![if !supportEmptyParas]>&nbsp;<![endif]><o:p></o:p></span></p><p class=MsoNormal><span style='font-family:Helvetica'>Additionally, we usejavascript, AJAX, and Web 2.0 design principles for designing easy-to-use,desktop-like web-based tools.<o:p></o:p></span></p><p class=MsoNormal><span style='font-family:Helvetica'><![if !supportEmptyParas]>&nbsp;<![endif]><o:p></o:p></span></p><p class=MsoNormal><span style='font-family:Helvetica'><b>How is the CoGegenomes database organized?</b></span><span style='font-family:Helvetica'><o:p></o:p></span></p><p class=MsoNormal><span style='font-family:Helvetica'><![if !supportEmptyParas]>&nbsp;<![endif]><o:p></o:p></span></p><a href = "picts/CoGe_database_core_schema.png"><img src = "picts/CoGe_database_core_schema.png" align = "right" width="200"></a><p class=MsoNormal><span style='font-family:Helvetica'>We have made greatefforts to keep the design of the database for storing and accessing genomicinformation simple while not compromising performance.<spanstyle="mso-spacerun: yes">&nbsp; </span>The core part of the database (thosetables that store genomic information specifically) consists of 13 tables.<spanstyle="mso-spacerun: yes">&nbsp; </span>The central concept around which allother information orbits is genomic "features".<span style="mso-spacerun:yes">&nbsp; </span>A genomic feature merely refers to some genomic region towhich additional information is attached.<span style="mso-spacerun: yes">&nbsp;</span>Take a "gene" as an example.<span style="mso-spacerun: yes">&nbsp;</span>Although the specific definition of a gene varies, abstractly it refersto a genomic region that contains biologically functional information.<spanstyle="mso-spacerun: yes">&nbsp; </span>This information, when read through aprocess called transcription, gives rise to the creation of RNAs.<spanstyle="mso-spacerun: yes">&nbsp; </span>These RNAs either carry out somecellular function (such as tRNAs and rRNAs) or encode information for use increating proteins through the process of translation.<span style="mso-spacerun:yes">&nbsp; </span>Thus, in order to track a region of a genome that hasassociated additional information, all we need to know is its location.<spanstyle="mso-spacerun: yes">&nbsp; </span>If you look at the schema for the CoGegenome database, you will see a relationship between the feature table and thelocation table.<span style="mso-spacerun: yes">&nbsp; </span>The feature tablestores only three types of information, the internal database id of a feature(each feature gets a unique id), the feature_type_id which links a feature tothe feature_type table (although you can probably guess the information storedin that table, we will go into more detail of the feature_type table further onin this discussion), and the data_information_id (which is the dataset fromwhich the feature was obtained).<span style="mso-spacerun: yes">&nbsp; </span>Thelocation table stores the information about the position of the feature:<spanstyle="mso-spacerun: yes">&nbsp; </span>the chromosome from which it isderived, the start position (nucleotide coordinates), the end position, thestrand (+/1 for top strand, -/-1 for bottom strand), and the feature databaseid (feature_id) to which the location belongs.<span style="mso-spacerun:yes">&nbsp; </span>The relationship between the feature and location tables isa "one to many" relationship where one feature can have more than one locationassociated to it.<span style="mso-spacerun: yes">&nbsp; </span>The reason forthis will become clear in a moment.<o:p></o:p></span></p><p class=MsoNormal><span style='font-family:Helvetica'><![if !supportEmptyParas]>&nbsp;<![endif]><o:p></o:p></span></p><p class=MsoNormal><span style='font-family:Helvetica'>Some genes containinformation for transcribing more than one RNA where different RNAs usedifferent genomic sequence contained within the gene.<span style="mso-spacerun:yes">&nbsp; </span>In order to track this kind of information (aka alternativesplicing), we need to specify additional genomic features for each RNA.<spanstyle="mso-spacerun: yes">&nbsp; </span>However, since not all of the DNAsequence of the gene is necessarily used for a given RNA, we may need to trackdifferent genomic locations for the gene and each RNA.<spanstyle="mso-spacerun: yes">&nbsp; </span>To do this, we can specify a uniquefeature for the gene and each RNA, and specify the appropriate genomiclocation(s) used for each.<span style="mso-spacerun: yes">&nbsp; </span>Byhaving a one to many relationship between the feature and location tables, wecan accommodate all this information.<span style="mso-spacerun: yes">&nbsp;</span>For example:<o:p></o:p></span></p><p class=MsoNormal><span style='font-family:Helvetica'><![if !supportEmptyParas]>&nbsp;<![endif]><o:p></o:p></span></p><p class=MsoNormal><span style='font-family:Helvetica'>GeneA:<spanstyle="mso-spacerun: yes">&nbsp; </span>Chromosome 1, Strand 1, Position (1000-10000)<o:p></o:p></span></p><p class=MsoNormal><span style='font-family:Helvetica'>RNA_A1: Chr 1, Strand 1,Position (1100-3000,4000-5000,8000,-9500)<o:p></o:p></span></p><p class=MsoNormal><span style='font-family:Helvetica'>RNA_A2: Chr 1, Strand 1,Position (1100-3000,4000-6000,7000-7500,8000,-9500)<o:p></o:p></span></p><p class=MsoNormal><span style='font-family:Helvetica'><![if !supportEmptyParas]>&nbsp;<![endif]><o:p></o:p></span></p><p class=MsoNormal><span style='font-family:Helvetica'>To extend this examplefurther, assume that each RNA listed above encodes a protein.<spanstyle="mso-spacerun: yes">&nbsp; </span>However, not all of the informationthat is transcribe into RNA is necessarily translated into protein.<spanstyle="mso-spacerun: yes">&nbsp; </span>In this case, we can specify additionalfeatures for just the protein coding region using the same process that was donefor the RNA coding regions.<span style="mso-spacerun: yes">&nbsp; </span><o:p></o:p></span></p><p class=MsoNormal><span style='font-family:Helvetica'><![if !supportEmptyParas]>&nbsp;<![endif]><o:p></o:p></span></p><p class=MsoNormal><span style='font-family:Helvetica'>One thing that you mayhave noticed from this example is that we specified several different kinds ofgenomic features:<span style="mso-spacerun: yes">&nbsp; </span>genes, RNAs andproteins coding regions (aka CDS).<span style="mso-spacerun: yes">&nbsp;</span>However, it is easy to find several more examples of different types ofgenomic features.<span style="mso-spacerun: yes">&nbsp; </span>For example,there are several different types of RNA genes besides those that code forproteins (aka mRNA):<span style="mso-spacerun: yes">&nbsp; </span>tRNA, rRNA,micro RNA, snRNA, snoRNA, etc.<span style="mso-spacerun: yes">&nbsp;</span>Additionally, there are several genomic features that don't code forRNA, but are useful to track because they have function. A TATA-box is such anexample of a genomic feature.<span style="mso-spacerun: yes">&nbsp;</span>TATA-boxes are genomic regions that contain the information needed toposition RNA polymerases (usually RNA pol II) correctly in order to transcribea gene in eukaryotic transcription systems.<o:p></o:p></span></p><p class=MsoNormal><span style='font-family:Helvetica'><![if !supportEmptyParas]>&nbsp;<![endif]><o:p></o:p></span></p><p class=MsoNormal><span style='font-family:Helvetica'>In order to track thiskind of information (different types of features), there is a table in the CoGegenomes database designed just for that purpose:<span style="mso-spacerun:yes">&nbsp; </span>the feature_type table.<span style="mso-spacerun:yes">&nbsp; </span>This table has three types of associated information:<spanstyle="mso-spacerun: yes">&nbsp; </span>the internal database id of the featuretype, the name of the feature type (e.g. mRNA), and the description of thefeature type (usually not used, but good to have just in case.)<o:p></o:p></span></p><p class=MsoNormal><span style='font-family:Helvetica'><![if !supportEmptyParas]>&nbsp;<![endif]><o:p></o:p></span></p><p class=MsoNormal><span style='font-family:Helvetica'>One problem with genomicfeatures is the plethora of alternative names that may be used to refer to asingle feature.<span style="mso-spacerun: yes">&nbsp; </span>To accommodatethis, there is a table in the database called feature_name.<spanstyle="mso-spacerun: yes">&nbsp; </span>Each entry in this table contains aname and description (optional) for a feature and is linked to the featuretable with a one-to-many relationship (i.e. one feature entry can have multiplefeature_name entries.)<o:p></o:p></span></p><p class=MsoNormal><span style='font-family:Helvetica'><![if !supportEmptyParas]>&nbsp;<![endif]><o:p></o:p></span></p><p class=MsoNormal><span style='font-family:Helvetica'>At this point, we havespecified the central concept of the CoGe genomes database: the genomicfeature.<span style="mso-spacerun: yes">&nbsp; </span>A feature needs to be ofa particular type (feature_type table), has genomic coordinates (locationtable), and one or more names (feature_name table).<span style="mso-spacerun:yes">&nbsp; </span>In addition to this information, there are additional kindsof information that can be associated with a feature.<span style="mso-spacerun:yes">&nbsp; </span>Generally, this additional information breaks into typetypes:<span style="mso-spacerun: yes">&nbsp; </span>specific sequences (e.g.translated protein sequence) and annotations.<span style="mso-spacerun:yes">&nbsp; </span>If a feature has a specific non-DNA sequence associated withit, such as a protein sequence, this information can be stored in the sequencetable.<span style="mso-spacerun: yes">&nbsp; </span>The sequence table in turnis associated with the sequence_type table to store the name and description ofthe type/kind of sequence.<span style="mso-spacerun: yes">&nbsp; </span><o:p></o:p></span></p><p class=MsoNormal><span style='font-family:Helvetica'><![if !supportEmptyParas]>&nbsp;<![endif]><o:p></o:p></span></p><p class=MsoNormal><span style='font-family:Helvetica'>The term "annotation" isa broad category of associated information about a gene that is gatheredthrough experimental and computational analysis of a genomic feature.<spanstyle="mso-spacerun: yes">&nbsp; </span>Some types of annotations arefunctional information, database references, literature references, cellularlocations, etc.<span style="mso-spacerun: yes">&nbsp; </span>In order to trackthis kind of information in a flexible yet organized manner, a three levelannotation schema is employed.<span style="mso-spacerun: yes">&nbsp; </span>Atthe root is the annotation table which stores an annotation, the feature towhich it belongs, and the type of annotation (stored in the annotation_typetable) that it is.<span style="mso-spacerun: yes">&nbsp; </span>In order toorganize types of annotations, there is an annotation_type table that has thename and description of the type of annotation.<span style="mso-spacerun:yes">&nbsp; </span>Additionally, annotation_types can be grouped into broadercategories using the annotation_type_group table.<span style="mso-spacerun:yes">&nbsp; </span>This tree-tiered approach allows maximum flexibility forspecifying annotations, types of annotations, and groups of types ofannotations.<span style="mso-spacerun: yes">&nbsp; </span><o:p></o:p></span></p><p class=MsoNormal><span style='font-family:Helvetica'><![if !supportEmptyParas]>&nbsp;<![endif]><o:p></o:p></span></p><p class=MsoNormal><span style='font-family:Helvetica'>An example of thisthree-tiered annotation approach is well suited for Geneontologyannotations.<span style="mso-spacerun: yes">&nbsp; </span>Let us say we havethe following annotation for a gene:<o:p></o:p></span></p><p class=MsoNormal><span style='font-family:Helvetica'><spanstyle="mso-spacerun: yes">&nbsp;</span>go_function: nucleic acid binding [goid0003676].<o:p></o:p></span></p><p class=MsoNormal><span style='font-family:Helvetica'><![if !supportEmptyParas]>&nbsp;<![endif]><o:p></o:p></span></p><p class=MsoNormal><span style='font-family:Helvetica'>For this annotation, wecan create an annotation_type_group with the name of "go_function", anannotation_type with the name of "0003676" and description "goid", and theannotation will be "nucleic acid binding".<span style="mso-spacerun:yes">&nbsp; </span>Using this type of a schema, it makes finding allannotation_types with an annotation_type_group of "go_function" trivial.<spanstyle="mso-spacerun: yes">&nbsp; </span>Likewise, it because easy to find allfeatures with an annotation_type of "0003676".<o:p></o:p></span></p><p class=MsoNormal><span style='font-family:Helvetica'><![if !supportEmptyParas]>&nbsp;<![endif]><o:p></o:p></span></p><p class=MsoNormal><span style='font-family:Helvetica'>The next major sectionof the CoGe database is for storing information about a dataset, the organismto which it belongs, and the actual genomic sequence.<span style="mso-spacerun:yes">&nbsp; </span>The key table for this information is the data_informationtable.<span style="mso-spacerun: yes">&nbsp; </span>Although this table should havebeen called the dataset table, the primary information this stores is the name,description, and version of a given dataset.<span style="mso-spacerun:yes">&nbsp; </span>Though it the data_source, feature, organism, andgenomic_sequence tables are linked.<span style="mso-spacerun: yes">&nbsp;</span>The data_source table stores the name, description and web-link of asource providing genomic data (e.g. genbank, JGI, TIGR, Ensembl).<spanstyle="mso-spacerun: yes">&nbsp; </span>The organism table stores the name anddescription of an organism (e.g. <i>Arabidopsis thaliana, Escherichia coli</i></span><spanstyle='font-family:Helvetica'>); the genomic_sequence table stores the genomicsequence.<span style="mso-spacerun: yes">&nbsp; </span>Since genomes can belarge (the human genome is ~3,000,000,000 base pairs), genomic sequence isbroken into chunks (usually 10,000 nucleotides).<o:p></o:p></span></p><p class=MsoNormal><span style='font-family:Helvetica'><![if !supportEmptyParas]>&nbsp;<![endif]><o:p></o:p></span></p><p class=MsoNormal><span style='font-family:Helvetica'>Using this schema tostore genomic data allows for storing genomes in any number ofconfigurations.<span style="mso-spacerun: yes">&nbsp; </span>For example, agenome (and all chromosomes contained therein) can be provided as a single fileor multiple files (e.g. one chromosome per file).<span style="mso-spacerun:yes">&nbsp; </span>Also, multiple versions of a genome can be stored byspecifying the "version" column of the data_information table.<spanstyle="mso-spacerun: yes">&nbsp; </span>And finally, additionalfeature/annotation files can be stored that are associated with a version of agenome, but do not actually contain genomic sequence (e.g. an annotation fileof all protein domains in a genome).<o:p></o:p></span></p><p class=MsoNormal><span style='font-family:Helvetica'><![if !supportEmptyParas]>&nbsp;<![endif]><o:p></o:p></span></p><p class=MsoNormal><span style='font-family:Helvetica'><b>How do I get a copyof the API, genome database, end-user tools, etc?<o:p></o:p></b></span></p><p class=MsoNormal><span style='font-family:Helvetica'><![if !supportEmptyParas]>&nbsp;<![endif]><o:p></o:p></span></p><p class=MsoNormal><span style='font-family:Helvetica'>As the CoGe system isstill in its early phases of development, we have not yet had time to finishthe documentation of the various components, package up the API, examplescripts, data files, etc. for general release.<span style="mso-spacerun:yes">&nbsp; </span>However, we would be more than happy to provide you withanything you might find helpful.<span style="mso-spacerun: yes">&nbsp;</span>Just send an e-mail to elyons(@)nature.berkeley.edu (remember to removethe "()" around the "@" symbol.)<span style="mso-spacerun: yes">&nbsp; </span><o:p></o:p></span></p></div><!-- ########################################## --><!-- ############END MAIN BODY################# --><!-- ########################################## --> </td>  <td  background="/CoGe/picts/table/table_bg_body_r.png">&nbsp;</td>  <td  background="/CoGe/picts/table/table_bg_head_r.png">&nbsp;</td> </tr> <tr>   <td height="9" valign="top" background="/CoGe/picts/table/table_bg_head_l.png"><img src="/CoGe/picts/table/spacer-1x1.png"></td>  <td height="9" valign="top"><img src="/CoGe/picts/table/table_bg_body_bl.png" width="6" height="9"></td>  <td height="9" valign="top" background="/CoGe/picts/table/table_bg_body_b.png"><img src="/CoGe/picts/table/spacer-1x1.png"></td>  <td height="9" valign="top"><img src="/CoGe/picts/table/table_bg_body_br.png" width="6" height="9"></td>  <td height="9" valign="top" width="6" background="/CoGe/picts/table/table_bg_head_r.png"><img src="/CoGe/picts/table/spacer-1x1.png"></td> </tr> <tr>   <td height="9" valign="bottom"><img src="/CoGe/picts/table/table_bg_head_bl.png" width="6" height="9"></td>  <td height="9" valign="bottom"><img src="/CoGe/picts/table/table_bg_head_b.png" width="6" height="9"></td>  <td height="9" valign="bottom" background="/CoGe/picts/table/table_bg_head_b.png"><img src="/CoGe/picts/table/spacer-1x1.png"></td>  <td height="9" valign="bottom"><img src="/CoGe/picts/table/table_bg_head_b.png" width="6" height="9"></td>  <td height="9" valign="bottom"><img src="/CoGe/picts/table/table_bg_head_br.png" width="6" height="9"></td> </tr></table> </body></html>